<a href="https://colab.research.google.com/github/diancastroherrera-dot/Exposiciones-multicapa/blob/main/Ejemplo_Red_multicapa_que_aprende_a_predecir_si_un_n%C3%BAmero_(0%E2%80%939)_es_mayor_o_menor_que_5_(EXPO).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Importar las librerías necesarias**
#equipo de Nallely y Eder

In [ ]:
import tensorflow as tf #ejecuta el modelo
from tensorflow import keras # simplifica la definición de las capas de la red
import numpy as np # crea y manipula los datos de entrada y salida


TensorFlow / Keras: permiten crear y entrenar redes neuronales fácilmente.

NumPy: se usa para manejar los datos como arreglos numéricos.


# **Crear los datos de ejemplo**


In [ ]:
# Datos de entrenamiento
X_data = np.array([[0], [1], [2], [3], [4], [5], [6], [7], [8], [9]], dtype=float) / 9
#datos de entrada
Y_labels = np.array([[0], [0], [0], [0], [0], [1], [1], [1], [1], [1]], dtype=float)
#respuestas correctas
# 0 = menor que 5, 1 = 5 o mayor

"X" son los números de entrada y los numeros que se van a clasificar.

Usamos dtype=float para compatibilidad con TensorFlow
Al dividir el array completo en 9 (el valor máximo en el conjunto de datos), se aplica la operación de división a cada número individualmente = (Escalamiento): Transforma el rango de los datos de [0, 9] al rango [0, 1]

"Y" son las respuestas correctas(salidas) dara 0 si es menor o igual a 5, 1 si es mayor a 5.
Esto es lo que la red va a aprender.


# **Red neuronal**

In [ ]:
# Definir el modelo (una red multicapa)
modelo = keras.Sequential([ #define el modelo con estructura secuencial y las capas se apilan una tras otra
    keras.layers.Dense(8, activation='relu', input_shape=[1]),#Capa Oculta 1: Recibe 1 entrada y
                                                              #procesa con 8 neuronas usando (activación ReLU)
    keras.layers.Dense(4, activation='relu'), #Capa Oculta 2: Procesa con 4 neuronas para extraer patrones (activación ReLU)
    keras.layers.Dense(1, activation='sigmoid')#Capa de Salida: 1 neurona con activación Sigmoid comprime
                                              # el resultado en un valor entre 0 y 1(probabilidad de clase)
])


Sequential: indica que las capas van una tras otra.

Dense: significa que cada neurona está conectada con todas las de la capa anterior.

activation='relu': ayuda a que la red aprenda patrones no lineales.

sigmoid: convierte la salida en un valor entre 0 y 1 (ideal para respuestas tipo “sí/no”).

Esta es la parte multicapa:
1 capa oculta con 8 neuronas,
1 capa oculta con 4 neuronas y
1 capa de salida con 1 neurona


# **Compilar el modelo (configurar cómo aprende)**


In [ ]:
# Configura el optimizador (Adam)
# y la función de pérdida (binary_crossentropy).

modelo.compile( # compila el modelo(prepara la red para el entrenamiento)
    optimizer='adam', #ajusta los pesos y segos de la red(reduce el error)
    loss='binary_crossentropy', #función de perdida(mide el error)
    metrics=['accuracy'] #metrica de evaluación
)

optimizer='adam': dice cómo ajustar los pesos y sesgos para aprender despues de cada época para reducir el error.


loss='binary_crossentropy': Mide el error entre la probabilidad de la red ($0.92$) y la respuesta real ($1$). El objetivo del entrenamiento es minimizar este valor..


metrics=['accuracy']: Métrica de Evaluación.	Le indica al usuario qué porcentaje de predicciones fueron correctas durante el entrenamiento.


# **Entrenamos el modelo y guardamos el historial**


In [ ]:
# El modelo itera 2000 veces (epochs) sobre los datos
#de entrenamiento para ajustar sus pesos.
history = modelo.fit(X_data, Y_labels, epochs=2000, verbose=0) #Entrenamiento Principal.
#Inicia el proceso de aprendizaje.

History: Guarda el registro completo de los valores de pérdida y precisión de cada una de las 2000 épocas.

fit: entrena la red con los datos x y y.

epochs=2000: repite el aprendizaje 2000 veces para mejorar.

verbose=0: oculta los mensajes de entrenamiento.


# **Probar la red con un número nuevo**


In [ ]:

pred_4 = modelo.predict(np.array([[4/9]]))
pred_8 = modelo.predict(np.array([[8/9]]))

print("Predicción para el número 4:", pred_4)
print("Predicción para el número 8:", pred_8)
# ... luego, el bloque final de print funcionaría con estas variables.

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 122ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 55ms/step
Predicción para el número 4: [[0.22117554]]
Predicción para el número 8: [[0.92634684]]


La red recibe un número y responde con un valor entre 0 y 1:


Cercano a 0 → menor o igual a 5


Cercano a 1 → mayor a 5


El código entrena un modelo que distingue entre dos categorías (clasificación binaria).

Muestra el flujo completo de una red multicapa (entrada → capas ocultas → salida).

Este último paso verifica si la red aprendió correctamente el patrón. El código ejecutará todo el proceso desde cero, entrenará un modelo para clasificar números, e imprimirá las probabilidades finales para el 4 y el 8

#Evaluamos la ultima precisión alcanzada

In [ ]:
#Extraccion de la metrica final
final_accuracy = history.history['accuracy'][-1]

la función modelo.fit(x, y, epochs=2000, ...) devuelve al finalizar las 2000 épocas de entrenamiento.

Función: history actúa como un registro completo de el proceso de aprendizaje del modelo. Almacena, época por época, cómo cambiaron las métricas y las pérdidas.

history: es un atributo dentro del objeto history y a su vez es un diccionario.

El diccionario .history utiliza los nombres de las métricas como sus claves (por ejemplo, 'loss', 'accuracy'). Los valores asociados a esas claves son listas que contienen el valor de la métrica en cada época.

-1: accede al último elemento de la lista y representa la precisión lograda en la última época de entrenamiento (la Época 2000)


# **Resultados**

In [ ]:
print("\n[METRICAS Y RESULTADOS FINALES]")
print(f"Precisión Final en Entrenamiento: {final_accuracy:.4f}")
print(f"Predicción para el número 4 (Esperado ≈ 0): {pred_4[0][0]:.4f}")
print(f"Predicción para el número 8 (Esperado ≈ 1): {pred_8[0][0]:.4f}")


[METRICAS Y RESULTADOS FINALES]
Precisión Final en Entrenamiento: 1.0000
Predicción para el número 4 (Esperado ≈ 0): 0.2212
Predicción para el número 8 (Esperado ≈ 1): 0.9263


# *CONCLUSIÓN *

El modelo alcanzó una precisión de entrenamiento perfecta ($1.0000$ o 100%), lo que confirma que, después de 2000 épocas, el optimizador Adam fue capaz de ajustar los pesos de la arquitectura 1-8-4-1 para mapear sin errores la regla lógica: $X \leq 5 \rightarrow 0$ y $X > 5 \rightarrow 1$.

Las pruebas de predicción verifican que la red ha aprendido a generalizar la regla a números desconocidos, utilizando el umbral de decisión de $0.5$:
Predicción para el número 8: El valor obtenido de $0.9263$ está firmemente por encima de $0.5$, lo que indica una clasificación correcta como Clase 1 (mayor a 5) con una confianza del 92.63%.Predicción para el número 4: El valor obtenido de $0.2212$ está claramente por debajo de $0.5$, lo que resulta en una clasificación correcta como Clase 0 (menor o igual a 5).


1–50 epocas:	Muy poco entrenamiento. La red no alcanza a aprender.	Predicciones ≈ 0.5 (aleatorias).

100–300	epocas: Aprende un poco, pero aún confunde valores cercanos al 5.	Precisión baja (~0.6–0.8).

500–1000 epocas: Comienza a distinguir correctamente.	Predicciones más cercanas a lo correcto (≈0 y ≈1).

2000 o más epocas:	La red converge completamente.	Precisión ≈ 1.0, predicciones muy acertadas.